Ballot comparison audits: Introduce Card Style Data math for more efficient sampling

nikhilb4a opened the pull request:
https://github.com/user-attachments/assets/98596f55-74fa-454a-a83e-925d874d8927

We are adding support for Card Style Data in ballot comparison audits (based on [this paper](https://dl.acm.org/doi/full/10.1145/3457907)). CSD allows for more efficient sampling by limiting the sample frame to cvrs that contain the contest, reducing the total sampling frame size and thus the sample size. 

This especially aids auditing contests that _do not_ appear on every ballot - with our previous approach, if a contest appears on 1/N cards, we would expect the required sample size to be N times larger, for the same risk limit/margin. With CSD, we restrict the sampling frame for each contest to only those CVRs that contain it. This means that for contests that appear on only a subset of ballots, we can still achieve the statistical confidence without inflating the overall sample size. So, we expect to sample approximately the same number of ballots containing a given contest—regardless of whether the contest appears on all cards or only a subset—for the same risk limit and margin.


## CSD has two key requirements
1. Use the partially diluted margin instead of the fully diluted margin in sample size calculating. The partially diluted margin = the margin/# **ballots with contest** while the fully diluted margin = the margin/# **total ballots, including those without contest**. This is implemented by updating how the `ballots` property is initiated, aggregating `metadata[contest.name]["total_ballots_cast"]` from each jurisdiction.
2. Limit the sampling pool to cvrs that include the contest. At sample time, we review the CVR interpretations to see if the contest is on the CVR.

jonahkagan commented on server/api/cvrs.py:
```diff
@@ -93,6 +94,8 @@ class CvrContestMetadata(TypedDict):
     # { choice_name: CvrChoiceMetadata }
     choices: Dict[str, CvrChoiceMetadata]
+    # { batch_id: List[int ballot_position] }
+    ballots_with_contest: Dict[str, List[int]]
```
> - likely still requires dealing with performance concern 1 when creating the relationships
> - replaces performance concern 2 with a ~similarly loaded query; instead of getting all ballot positions after they are assigned, we will create contest relationships for each ballot

I think we could get around this since the CVRs are processed in a streaming manner, so each ballot can have its contests derived and written to the database while it's being processed.

> requires an additional query at sampling time to build the manifest from CvrBallots

We'll be able to query just the CVRs needed for the contest we're sampling for, so I think this will potentially be performant given the right database indexes, but we'll definitely need to test it. In the worst case, it will still probably be faster than loading all the CVR ballots for an election and filtering them in memory, which is what the PR does currently.

I'll pass down some wisdom that @benadida has hammered into my brain after many technical discussions similar to this one, which is that it's almost always better to use the database if you're worried about performant manipulation of large data sets, since that's exactly what it's designed to be good at.

nikhilb4a replied:
> I think we could get around this since the CVRs are processed in a streaming manner, so each ballot can have its contests derived and written to the database while it's being processed.

Ah okay, nice. My concern was that since the current approach processes each CVR as a CvrBallot by writing it to a temp file and doesn't do any other db actions, I assumed we'd have to wait till that finishes, keep a cache of the ballot-contest names, and then create the relationships after the fact in a similar way to how BallotPositions are added after the fact. But, I realize now (what you probably already were aware of) we can follow the same approach we take for CvrBallot, and write the relationships to a temp file CvrBallotContest which we then Copy over. Nice, that'll be efficient

> We'll be able to query just the CVRs needed for the contest we're sampling for, so I think this will potentially be performant given the right database indexes, but we'll definitely need to test it. In the worst case, it will still probably be faster than loading all the CVR ballots for an election and filtering them in memory, which is what the PR does currently.

Okay, that seems to make sense, and good point on indexes. I'm still wrapping my head around the last sentence, could I clarify – The current PR doesn't do the filtering at sample time, but aggregates the pre-filtered `ballots_with_contest` from each jurisdiction at sample time, indexed on the contest so it is only grabbing ballot positions with the contest for each jurisdiction, which I would expect to be pretty fast. **But**, now that I write that, I see maybe you are pointing out that loading the contest metadata into memory is actually the bottleneck here? Maybe my mental model was off there not considering how slow that would be if `contest_metadata` gets really large with the jurisdiction's whole map of contest_names to ballot_positions. After a couple gpt queries, I see the sqlite indexed query might be ~1ms and loading the potentially ~50mb "ballots_with_contests" object  (1 million cvrs) might be ~0.5s, per jurisdiction, supporting your point clearly.

> I'll pass down some wisdom that @benadida has hammered into my brain after many technical discussions similar to this one, which is that it's almost always better to use the database if you're worried about performant manipulation of large data sets, since that's exactly what it's designed to be good at

That's great advice, thanks for sharing it! I appreciate you talking through these points with me, really helpful, and I am in agreement that the db approach is the right path.

nikhilb4a replied:
Lol well it led me down a helpful path!

jonahkagan commented on server/api/contests.py:
```diff
@@ -281,8 +281,11 @@ def validate_contests(contests: List[JSONDict], election: Election):
 def set_contest_metadata(election: Election):
     for contest in election.contests:
-        if election.audit_type != AuditType.BALLOT_POLLING:
+        if election.audit_math_type == AuditMathType.CARDSTYLEDATA:
```
Nit: Nest this under the ballot comparison case to clarify that this is one of the two options within that audit type

jonahkagan commented on server/models.py:
```diff
@@ -1013,13 +1021,43 @@ class CvrBallot(Base):
+            ["cvr_ballot.batch_id", "cvr_ballot.record_id"],
+            ondelete="cascade",
+        ),
+    )
```
This database model looks great! I'm definitely thinking about how large this table will grow to be if we start using CSD more often, but I think we can cross that bridge when we get to it. Have you tested with at least the scale of the upcoming pilot election as a sanity check?

jonahkagan reviewed (commented):
This looks great overall, pending confirmation that the sampling logic is correct.

My only blocking concern is the question about changing the contests after uploading the CVRs.

nikhilb4a commented on server/api/cvrs.py:
```diff
@@ -1393,6 +1419,17 @@ def parse_cvrs():
                     contests_metadata[contest_name]["total_ballots_cast"] += 1
+                    if should_save_cvr_ballot_contests and ballot_contests_csv:
+                        contest_id = contest_name_to_id.get(contest_name)
+                        # If contest_id is None, the contest is not being audited
```
Good point, I had a misunderstanding there about the possible sequencing of actions. I'll think about this some more, with two initial wonderings being 1) do we recompute CvrBallotContests when the audited contests change or 2) do we use name instead of id when building CvrBallotContests (though this would break the relation).. I'll give this some thought but open to advice

nikhilb4a commented on server/api/cvrs.py:
```diff
@@ -205,6 +205,19 @@ def cvr_contests_metadata(
+    for jurisdiction in contest.jurisdictions:
+        metadata = cvr_contests_metadata(jurisdiction)
+        assert metadata is not None
+        total_ballots += metadata[contest.name]["total_ballots_cast"]
```
Had the same thought, though tabling for now as we go with the non `CvrBallotContest` approach.

nikhilb4a commented on server/api/rounds.py:
```diff
@@ -495,7 +495,7 @@ def validate_sample_size(round: dict, election: Election):
                 ["asn", "0.9", "0.8", "0.7", "custom", "all-ballots"]
             ),
-            AuditType.BALLOT_COMPARISON: ["supersimple", "custom"],
+            AuditType.BALLOT_COMPARISON: ["supersimple", "custom", "cardstyledata"],
```
Will do. Thinking I should still support "supersimple" in the `autoselect_sample_size` to make sure we are backwards compatible with existing audits, though just want to sanity check that is needed

nikhilb4a commented on server/tests/ballot_comparison/test_ballot_comparison_cardstyledata.py:
```diff
@@ -0,0 +1,386 @@
+    [{"audit_math_type": AuditMathType.CARDSTYLEDATA}],
+    indirect=True,
+)
+def test_ballot_comparison_cardstyledata_two_rounds(
```
Added logic to test for deleting/reuploading cvrs to check that `total_ballots_cast` is cleared, since that is different in this flow vs other ballot comparisons

nikhilb4a commented:
Updated manifest creation logic to parse `CvrBallot.interpretations` at sample time, dropped `CvrBallotContest` file. This helps CVR upload speed stay the same as before. Sample time is slower, but it's on the order of seconds (1 million CVRs processes in 5s vs 3s).
